### PR TITLE
Network types documentation

### PR DESCRIPTION
**Pull request by ru-fu**
Update the documentation pages about network types to make them consistent and up-to-date.

### CODE REVIEW

File: doc/reference/network_bridge.md [hunk 1]
```diff
@@ -6,39 +6,50 @@ discourse: 7322
 # Bridge network
 
 As one of the possible network configuration types under LXD, LXD supports creating and managing network bridges.
+<!-- Include start bridge intro -->
+A network bridge creates a virtual L2 ethernet switch that instance NICs can connect to, making it possible for them to communicate with each other and the host.
 LXD bridges can leverage underlying native Linux bridges and Open vSwitch.
```
Comment (ru-fu) [inline doc/reference/network_bridge.md:11]: This line is from the original documentation (which is probably from before OVN support) - is it still correct?

File: doc/reference/network_external.md [hunk 2]
```diff
@@ -1,5 +1,15 @@
 # External networks
 
+<!-- Include start external intro -->
+External networks use network interfaces that already exist.
+Therefore, LXD has limited possibility to control them, and LXD features like network ACLs, network forwards and network zones are not supported.
+
+The main purpose for using external networks is to provide an uplink network through a parent interface.
+This external network specifies the presets to use when connecting instances to a parent interface.
+
+LXD supports the following external network types:
```
Comment (ru-fu) [inline doc/reference/network_external.md:10]: This is slightly extended from what we had in the networks overview originally - please check.

File: doc/reference/network_ovn.md [hunk 3]
```diff
@@ -5,44 +5,78 @@ discourse: 11033
...
 (network-ovn-options)=
 ## Configuration options
 
+The following configuration key namespaces are currently supported for the `ovn` network type:
+
+ - `bridge` (L2 interface configuration)
+ - `dns` (DNS server and resolution configuration)
+ - `ipv4` (L3 IPv4 configuration)
+ - `ipv6` (L3 IPv6 configuration)
+ - `security` (network ACL configuration)
```
Comment (ru-fu) [inline doc/reference/network_ovn.md:31]: Is it correct that OVN doesn't support maas?
Before, the docs said that all network types support it.

File: doc/reference/network_physical.md [hunk 4]
```diff
@@ -1,30 +1,58 @@
...
+{{note_ip_addresses_CIDR}}
+```
+
+The following configuration options are available for the `physical` network type:
+
 Key                             | Type      | Condition             | Default                   | Description
 :--                             | :--       | :--                   | :--                       | :--
+gvrp                            | boolean   | -                     | false                     | Register VLAN using GARP VLAN Registration Protocol
+mtu                             | integer   | -                     | -                         | The MTU of the new interface
+parent                          | string    | -                     | -                         | Parent interface to create `physical` NICs on
```
Comment (ru-fu) [inline doc/reference/network_physical.md:34]: This said "sriov NICs" before - correct to change it?

File: doc/reference/network_bridge.md (hunk 1 above)
Comment (tomponline) [reply doc/reference/network_bridge.md:11]: This is still correct AFAIK, the Open vSwitch reference is still valid (it can create or use an Open vSwitch bridge) and is distinct from OVN networks.

File: doc/reference/network_physical.md (hunk 4 above)
Comment (tomponline) [reply doc/reference/network_physical.md:34]: Ah yes it was wrong before, but it should be:

> Existing interface to use for network

Rather than indicating that it is only used for `physical` NICs (it could only be used for 1 physical NIC at a time anyway) and it can be used as the uplink for OVN networks too.

File: doc/reference/network_ovn.md (hunk 3 above)
Comment (tomponline) [reply doc/reference/network_ovn.md:31]: Yes it cannot register the subnet in MAAS as it doesn't use a MAAS subnet, but rather its own private overlay subnet.

File: doc/reference/network_bridge.md [hunk 5]
```diff
@@ -6,39 +6,50 @@ discourse: 7322
...
+<!-- Include end bridge intro -->
 
-Creation and management of LXD bridges is performed via the `lxc network` command.
-A bridge created by LXD is by default "managed" which means that LXD also will additionally set up a local `dnsmasq`
-DHCP server and if desired also perform NAT for the bridge (this is the default.)
+The `bridge` network type allows to create an L2 bridge that connects the instances that use it together into a single network L2 segment.
+A bridge created by LXD is by default "managed", which means that in addition to the bridge, LXD sets up a local `dnsmasq` DHCP server.
```
Comment (tomponline) [inline doc/reference/network_bridge.md:15]: The use of "by default" here suggests a LXD bridge can be something other than "managed", which it cannot.

Comment (tomponline) [reply doc/reference/network_bridge.md:15]: How about instead:

> LXD created bridges are managed, which means that in addition to the creating the bridge interface itself, LXD also sets up a local `dnsmasq` process to provide DHCP, IPv6 route announcements and DNS services to the network.

File: doc/reference/network_ovn.md [hunk 6]
```diff
@@ -5,44 +5,78 @@ discourse: 11033
...
-A LXD OVN network can be connected to an existing managed LXD bridge network in order for it to gain outbound
-access to the wider network. All connections from the OVN logical networks are NATed to a dynamic IP allocated by
-the parent network.
+The `ovn` network type allows to create logical networks using the OVN {abbr}`SDN (software-defined networking)`.
+This kind of network can be useful for labs and multi-tenant environments where the same logical subnets are used in multiple discrete networks.
 
-```{toctree}
-:maxdepth: 1
+A LXD OVN network can be connected to an existing managed {ref}`network-bridge` or {ref}`network-physical` to gain outbound access to the wider network.
+All connections from the OVN logical networks are NATed to a dynamic IP allocated by the parent network.
```
Comment (tomponline) [inline doc/reference/network_ovn.md:18]: > By default all connections from the OVN logical networks are NATed to an IP allocated from the uplink network.

File: doc/reference/network_sriov.md [hunk 7]
```diff
@@ -1,17 +1,32 @@
 (network-sriov)=
 # SR-IOV network
 
-The sriov network type allows one to specify presets to use when connecting instances to a parent interface
-using sriov NICs. This allows the instance NIC itself to simply specify the `network` it is connecting to without
-knowing any of the underlying configuration details.
+<!-- Include start SR-IOV intro -->
+{abbr}`SR-IOV (Single root I/O virtualization)` is a hardware standard that allows a single network interface to appear as several virtual network interfaces in a virtualized environment.
```
Comment (tomponline) [inline doc/reference/network_sriov.md:5]: Here lets change `single network interface` to `single network card port` as in this case it is infact a physical mapping of ports to virtual interfaces.

File: doc/explanation/networks.md [hunk 8]
```diff
@@ -49,7 +49,11 @@ Fully controlled networks create network interfaces and provide most functionali
 LXD supports the following network types:
 
 {ref}`network-bridge`
-: A network bridge creates a virtual L2 ethernet switch that instance NICs can connect to, making it possible for them to communicate with each other and the host.
+: % Include content from [../reference/network_bridge.md](../reference/network_bridge.md)
+  ```{include} ../reference/network_bridge.md
+      :start-after: <!-- Include start bridge intro -->
+      :end-before: <!-- Include end bridge intro -->
+  ```
```
Comment (stgraber) [inline doc/explanation/networks.md:56]: Make one commit which has those kind of changes (removes text replaces by matching include)

File: doc/reference/network_bridge.md [hunk 9]
```diff
@@ -6,39 +6,50 @@ discourse: 7322
 # Bridge network
 
 As one of the possible network configuration types under LXD, LXD supports creating and managing network bridges.
+<!-- Include start bridge intro -->
+A network bridge creates a virtual L2 ethernet switch that instance NICs can connect to, making it possible for them to communicate with each other and the host.
 LXD bridges can leverage underlying native Linux bridges and Open vSwitch.
+<!-- Include end bridge intro -->
```
Comment (stgraber) [inline doc/reference/network_bridge.md:12]: Put those in the same commit as the one adding the content include and removing the text (previous comment).

File: doc/reference/network_bridge.md [hunk 10]
```diff
@@ -6,39 +6,50 @@ discourse: 7322
...
 
-Creation and management of LXD bridges is performed via the `lxc network` command.
-A bridge created by LXD is by default "managed" which means that LXD also will additionally set up a local `dnsmasq`
-DHCP server and if desired also perform NAT for the bridge (this is the default.)
+The `bridge` network type allows to create an L2 bridge that connects the instances that use it together into a single network L2 segment.
+Bridges created by LXD are managed, which means that in addition to creating the bridge interface itself, LXD also sets up a local `dnsmasq` process to provide DHCP, IPv6 route announcements and DNS services to the network.
+By default, it also performs NAT for the bridge.
 
-When a bridge is managed by LXD, configuration values under the `bridge` namespace can be used to configure it.
+See {ref}`network-bridge-firewalld` for instructions on how to configure firewalld to work with LXD bridge networks.
```
Comment (stgraber) [inline doc/reference/network_bridge.md:18]: Move those general content edits to own commit, one commit per page.

File: doc/reference/network_bridge.md [hunk 11]
```diff
@@ -6,39 +6,50 @@ discourse: 7322
...
 
-Integrate with systemd-resolved </howto/network_bridge_resolved>
-Configure Firewalld </howto/network_bridge_firewalld>
-```
+If you're using IPv6 for your bridge network, you should use a prefix size of 64.
+
+Larger subnets (i.e., using a prefix smaller than 64) should work properly too, but they aren't typically that useful for {abbr}`SLAAC (Stateless Address Auto-configuration)`.
+
+Smaller subnets are in theory possible (when using stateful DHCPv6 for IPv6 allocation), but they aren't properly supported by `dnsmasq` and might cause problems.
+If you must create a smaller subnet, use static allocation or another standalone router advertisement daemon.
```
Comment (stgraber) [inline doc/reference/network_bridge.md:27]: Assuming this got moved without edits, make the move its own commit. If it got moved and edited, commit the move first (with no edits) and then include the edits as part of the content edit commit for the page.

File: doc/reference/network_bridge.md [hunk 12]
```diff
@@ -6,39 +6,50 @@ discourse: 7322
...
+```{note}
+{{note_ip_addresses_CIDR}}
+An exception are tunnel local and remote addresses.
+These should be specified as plain addresses (`192.0.2.0` or `2001:db8::`).
+```
 
-The exception being tunnel local and remote addresses which are just plain addresses (`1.1.1.1` or `fd80:1234::1`).
+The following configuration options are available for the `bridge` network type:
 
 Key                                  | Type      | Condition             | Default                   | Description
```
Comment (stgraber) [inline doc/reference/network_bridge.md:54]: The kind of small consistency edits as seen in the table here, could do with their own commit as that requires less active review (it's a mostly mechanical change).

File: doc/substitutions.yaml [hunk 13]
```diff
@@ -1,2 +1,2 @@
 # Key/value substitutions to use within the Sphinx doc.
-{}
+{note_ip_addresses_CIDR: "IP addresses and subnets must be specified in CIDR notation (for example, `192.0.2.0/24` or `2001:db8::/32`)."}
```
Comment (stgraber) [inline doc/substitutions.yaml:2]: Make the new substitution definition its own commit.

**Review by stgraber — CHANGES_REQUESTED** [review]
The end result is good but let's use this PR to sort out how we split commits for a change like this so we end up with something that's easier to review and backport.

**Comment by stgraber** [comment]
So overall for this PR, I think I'd expect (might have missed some):
 - doc/network: Alphabetical ordering of tables
 - doc/network: Add ID for easier linking
 - doc/substitutions: Add note_ip_addresses_CIDR
 - doc/network/XYZ: Table consistency
 - doc/network: Extend introduction definition
 - doc/networks: Include network driver definitions
 - doc/network/bridge: Move IPv6 prefix section
 - doc/network/XYZ: Update content

With both of the XYZ entries actually being multiple commits, one per page.
This should make it significantly easier to review commit by commit as it clearly separates the mechanical changes from the rest and lets us better focus on what needs closer attention during review.

**Comment by ru-fu** [comment]
I'm not sure I agree that this number of commits make it easier to review (especially since there was quite a bit of outdated information in these topics that can only be caught when reviewing in context), but updated.

File: doc/reference/network_physical.md [hunk 14]
```diff
@@ -11,24 +11,27 @@ This allows the instance NIC itself to simply specify the `network` it is connec
...
-ipv6.routes.anycast             | boolean   | ipv6 address          | false                     | Allow the overlapping routes to be used on multiple networks/NIC at the same time.
-maas.subnet.ipv4                | string    | ipv4 address          | -                         | MAAS IPv4 subnet to register instances in (when using `network` property on nic)
-maas.subnet.ipv6                | string    | ipv6 address          | -                         | MAAS IPv6 subnet to register instances in (when using `network` property on nic)
-ovn.ingress\_mode               | string    | standard mode         | l2proxy                   | Sets the method that OVN NIC external IPs will be advertised on uplink network. Either `l2proxy` (proxy ARP/NDP) or `routed`.
+dns.nameservers                 | string    | standard mode         | -                         | List of DNS server IPs on `physical` network
+ipv4.gateway                    | string    | standard mode         | -                         | IPv4 address for the gateway and network
```
Comment (stgraber) [inline doc/reference/network_physical.md:40]: Why did we drop the mention of CIDR notation?

As it's written now, I would put in `192.168.0.1` and then get an error telling me it's not a valid address because LXD expects `192.168.0.1/24`.

File: doc/reference/network_physical.md [hunk 15]
```diff
@@ -11,24 +11,27 @@ This allows the instance NIC itself to simply specify the `network` it is connec
...
+ipv4.ovn.ranges                 | string    | -                     | -                         | Comma-separated list of IPv4 ranges to use for child OVN network routers (FIRST-LAST format)
+ipv4.routes                     | string    | ipv4 address          | -                         | Comma-separated list of additional IPv4 CIDR subnets that can be used with child OVN networks `ipv4.routes.external` setting
+ipv4.routes.anycast             | boolean   | ipv4 address          | false                     | Allow the overlapping routes to be used on multiple networks/NIC at the same time
+ipv6.gateway                    | string    | standard mode         | -                         | IPv6 address for the gateway and network
```
Comment (stgraber) [inline doc/reference/network_physical.md:44]: Same wrt CIDR

File: doc/substitutions.yaml (hunk 13 above)
Comment (stgraber) [inline doc/substitutions.yaml:2]: I think we do need it on a per-option basis because it's not always true.

For example `ipv4.address` on a NIC device isn't CIDR but entries in `ipv4.routes` or `ipv4.routes.external` are CIDR.
On a network, `ipv4.address` and `ipv4.gateway` are CIDR but `ipv4.nat.address` is not.

File: doc/reference/network_bridge.md [hunk 16]
```diff
@@ -37,21 +37,25 @@ another standalone RA daemon be used.
...
  - `raw` (raw configuration file content)
+ - `tunnel` (cross-host tunneling configuration)
+ - `user` (free-form key/value for user metadata)
 
-It is expected that IP addresses and subnets are given using CIDR notation (`1.1.1.1/24` or `fd80:1234::1/64`).
-
-The exception being tunnel local and remote addresses which are just plain addresses (`1.1.1.1` or `fd80:1234::1`).
+```{note}
+{{note_ip_addresses_CIDR}}
+An exception are tunnel local and remote addresses.
```
Comment (stgraber) [inline doc/reference/network_bridge.md:48]: Ah yeah, forgot those two on top of my list in the previous commit ;)

File: doc/reference/network_bridge.md [hunk 17]
```diff
@@ -6,26 +6,25 @@ discourse: 7322
 # Bridge network
 
 As one of the possible network configuration types under LXD, LXD supports creating and managing network bridges.
-L<!-- Include start bridge intro -->
```
Comment (stgraber) [inline doc/reference/network_bridge.md]: When was that typo introduced? If it was an earlier commit in this branch, it'd be better to fix that earlier commit to make this one's diff less confusing (and less likely to conflict on cherry-picks).

**Comment by stgraber** [comment]
Thanks, that's significantly easier to go through!

So remaining issues are the commit titles which need adjustment to use prefixes and some tweaks to the CIDR stuff.

I think the substitution should be used for something like:
```
Wherever network subnet information is required, LXD uses the [CIDR notation](https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing), for example, `192.0.2.0/24` or `2001:db8::/32`.

This does not apply to cases where a single address is required, for example local/remote addresses of tunnels, NAT address or specific addresses to apply to an instance.
```

But then the tables should still say `(CIDR)` next to anything which does expect an address+subnet info.
